### PR TITLE
python3Packages.miasm: init at 0.1.5-unstable-2024-04-28

### PR DESCRIPTION
**Pull request by msanft**
## Description of changes

This adds the [miasm](https://github.com/cea-sec/miasm?tab=readme-ov-file) Python reverse engineering framework.

Related to #81418 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/miasm/default.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+, llvmlite
+, parameterized
+, pycparser
+, pyparsing
+, z3-solver
+}:
+buildPythonPackage rec {
+  pname = "miasm";
+  version = "0.1.3-unstable-2024-04-28";
+  format = "setuptools";
```
Comment (fabaff) [inline pkgs/development/python-modules/miasm/default.nix:15]:
```suggestion
  pyproject = true";
```

and then `  build-system = [ setuptools ];`.

File: pkgs/development/python-modules/miasm/default.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    pyparsing
+    z3-solver
+  ];
+
+  dependencies = [ gcc ];
+
+  pythonImportsCheck = [ "miasm" ];
+
+  meta = {
+    description = "Reverse engineering framework in Python ";
```
Comment (fabaff) [inline pkgs/development/python-modules/miasm/default.nix:46]:
```suggestion
    description = "Reverse engineering framework in Python";
```

File: pkgs/development/python-modules/miasm/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    homepage = "https://github.com/cea-sec/miasm";
```
Comment (fabaff) [inline pkgs/development/python-modules/miasm/default.nix:47]:
```suggestion
    homepage = "https://github.com/cea-sec/miasm";
    changelog = "https://github.com/cea-sec/miasm/blob/v${version}/CHANGELOG.md";
```

File: pkgs/development/python-modules/miasm/default.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    license = lib.licenses.gpl2;
```
Comment (fabaff) [inline pkgs/development/python-modules/miasm/default.nix:48]: According to a couple of file headers (e.g., https://github.com/cea-sec/miasm/blob/master/miasm/os_dep/win_api_x86_32_seh.py), the license is GPLv2+.

```suggestion
    license = lib.licenses.gpl2Plus;
```

**Review by fabaff — COMMENTED** [review]
Upstream has a tag for [0.1.5](https://github.com/cea-sec/miasm/releases/tag/v0.1.5).

One soon to become issue is that upstream still depends on `distutils` which will be removed with Python 3.12.

File: pkgs/development/python-modules/miasm/default.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  patches = [
+    #  Use a valid semver as now required by setuptools
+    ./0001-setup.py-use-valid-semver.patch
+
+    # Removes the (unfree) IDAPython dependency
+    ./0002-core-remove-IDAPython-dependency.patch
+  ];
+
+  propagatedBuildInputs = [
```
Comment (fabaff) [inline pkgs/development/python-modules/miasm/default.nix:32]:
```suggestion
  dependencies = [
```

File: pkgs/development/python-modules/miasm/default.nix [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    future
+    llvmlite
+    parameterized
+    pycparser
+    pyparsing
+    z3-solver
+  ];
+
+  dependencies = [ gcc ];
```
Comment (fabaff) [inline pkgs/development/python-modules/miasm/default.nix:41]: This smells like a `buildInput` at first glance.

**Comment by msanft** [comment]
> One soon to become issue is that upstream still depends on distutils which will be removed with Python 3.12.

Valid point. As the repo is more or less actively maintained, we might get them to update it for 3.12 support. A patchset would be quite heavy I think.

**Pull request merged**